### PR TITLE
cudaMallocAsync: If OOM, sync the stream and retry allocation.

### PR DESCRIPTION
**Pull request by nouiz**
The sync allow the driver more option to find memory. So sometimes it can find memory available after a sync.

@sanjoy

### CODE REVIEW

File: tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc [hunk 1]
```diff
@@ -207,9 +207,16 @@ void* GpuCudaMallocAsyncAllocator::AllocateRaw(size_t alignment,
   }
   se::cuda::ScopedActivateExecutorContext scoped_activation{stream_exec_};
   void* ptr = nullptr;
-  if (auto result =
-          cuMemAllocFromPoolAsync(reinterpret_cast<CUdeviceptr*>(&ptr),
-                                  num_bytes, pool_, cuda_stream_)) {
+  auto result = cuMemAllocFromPoolAsync(reinterpret_cast<CUdeviceptr*>(&ptr),
+                                        num_bytes, pool_, cuda_stream_);
+  if (result == CUDA_ERROR_OUT_OF_MEMORY) {
+    // This can fix some OOM when memory is tight.
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc:213]: Can you please add some more detail on the situations in which this helps?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc (hunk 1 above)
Comment (nouiz) [reply tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc]: done.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Can this theoretically introduce a deadlock?  E.g. imagine that the stream has a NCCL all-reduce operation on it, and the matching all-reduce (on the other device) was going to be launched after this allocation returned.  Synchronizing on the stream would deadlock in this case.

WDYT about synchronizing with a timeout?

**Comment by nouiz** [comment]
> Can this theoretically introduce a deadlock? E.g. imagine that the stream has a NCCL all-reduce operation on it, and the matching all-reduce (on the other device) was going to be launched after this allocation returned. Synchronizing on the stream would deadlock in this case.

I checked with the NCCL team and it would be safe for some version only.

> WDYT about synchronizing with a timeout?

I'll do that to have a version that is save for all NCCL version and maybe other potential deadlock.

**Comment by nouiz** [comment]
Maybe just a clarification. It can deadlock only because in corner case, we can have only 1 threads per device that does the dispatch. If it was sure that we would always have 1 dispatch thread per GPU, there wouldn't be any risk of deadlock as all the nccl call would be done.

I specify this as if TFRT could make this guarantee, it would bypass the current problem.

**Comment by nouiz** [comment]
Sadly just doing a way doesn't trigger what we want in the driver. So I'll close this PR.
Thanks for the review.

**Pull request closed without merging**